src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexer.cs

```diff
@@ -0,0 +1,298 @@
+
+        private const string DbKey = "AddrData";
+
+        private SubscriptionToken blockConnectedSubscription, blockDisconnectedSubscription;
+
+        /// <summary>Queue that is populated when block is connected or disconnected.</summary>
+        /// <remarks><c>bool</c> key is <c>true</c> when block is connected.</remarks>
+        private AsyncQueue<KeyValuePair<bool, ChainedHeaderBlock>> blockReceivedQueue;
+
+        private LiteDatabase db;
+
+        private LiteCollection<AddressIndexerData> dataStore;
+
+        private AddressIndexerData addressesIndex;
+
+        public AddressIndexer(StoreSettings storeSettings, ISignals signals, DataFolder dataFolder, ILoggerFactory loggerFactory,
+            Network network, IBlockStore blockStore, INodeStats nodeStats, IConsensusManager consensusManager)
+        {
+            this.signals = signals;
+            this.storeSettings = storeSettings;
+            this.network = network;
+            this.blockStore = blockStore;
+            this.nodeStats = nodeStats;
+            this.dataFolder = dataFolder;
+            this.consensusManager = consensusManager;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        public void Initialize()
+        {
+            if (!this.storeSettings.TxIndex || !this.storeSettings.IndexAddresses)
+            {
+                this.logger.LogTrace("(-)[DISABLED]");
+                return;
+            }
+
+            string dbPath = Path.Combine(this.dataFolder.RootPath, "addressindex.litedb");
+            this.db = new LiteDatabase(new ConnectionString() {Filename = dbPath});
+
+            this.logger.LogDebug("TxIndexing is enabled.");
+
+            this.dataStore = this.db.GetCollection<AddressIndexerData>(DbKey);
+
+            this.addressesIndex = this.dataStore.FindAll().FirstOrDefault();
+
+            if (this.addressesIndex == null)
+            {
+                this.logger.LogDebug("Tip was not found, initializing with genesis.");
+
+                this.addressesIndex = new AddressIndexerData()
+                {
+                    TipHash = this.network.GenesisHash.ToString(),
+                    AddressIndexDatas = new List<AddressIndexData>()
+                };
+                this.dataStore.Insert(this.addressesIndex);
+            }
+
+            this.blockReceivedQueue = new AsyncQueue<KeyValuePair<bool, ChainedHeaderBlock>>(this.OnEnqueueAsync);
+
+            // Subscribe to events.
+            this.blockConnectedSubscription = this.signals.Subscribe<BlockConnected>(blockConnectedData =>
+            {
+                while (this.blockReceivedQueue.Count > 100)
+                {
+                    this.logger.LogWarning("Address indexing is slowing down the consensus.");
+                    Thread.Sleep(5000);
+                }
+
+                this.blockReceivedQueue.Enqueue(new KeyValuePair<bool, ChainedHeaderBlock>(true, blockConnectedData.ConnectedBlock));
```

Comment:
I wonder if you may want to take the Block instance out here, it will disappear from `blockConnectedData` if more then 100 blocks behind consensus (alternatively you can fetch it form disk later) ok to do it in a following PR.